libusb/os/linux_android_jni.c

```diff
@@ -0,0 +1,1052 @@
+		free(*jni);
+		return r;
+	}
+
+	r = android_jni_fill_ctx_ids(*jni, jni_env);
+	if (r != LIBUSB_SUCCESS) {
+		free(*jni);
+		return r;
+	}
+
+	/* ActivityThread activity_thread = ActivityThread.currentActivityThread(); */
+	ActivityThread = (*jni_env)->FindClass(jni_env, "android/app/ActivityThread");
+	ActivityThread__currentActivityThread =
+		(*jni_env)->GetStaticMethodID(jni_env,
+			ActivityThread, "currentActivityThread",
+			"()Landroid/app/ActivityThread;");
+	ActivityThread_getApplication =
+		(*jni_env)->GetMethodID(jni_env,
+			ActivityThread, "getApplication", "()Landroid/app/Application;");
+	activity_thread =
+		(*jni_env)->CallStaticObjectMethod(jni_env,
+			ActivityThread, ActivityThread__currentActivityThread);
+
+	/* Context local_context = activity_thread.getApplication(); */
+	local_context =
+		(*jni_env)->CallObjectMethod(jni_env,
+			activity_thread, ActivityThread_getApplication);
+
+	(*jni)->application_context = (*jni_env)->NewGlobalRef(jni_env, local_context);
+
+
+	Context = (*jni_env)->FindClass(jni_env, "android/content/Context");
+	Context__USB_SERVICE =
+		(*jni_env)->GetStaticObjectField(jni_env,
+			Context,
+			(*jni_env)->GetStaticFieldID(jni_env,
+				Context, "USB_SERVICE", "Ljava/lang/String;"));
+	Context_getPackageManager =
+		(*jni_env)->GetMethodID(jni_env,
+			Context, "getPackageManager",
+			"()Landroid/content/pm/PackageManager;");
+	Context_getSystemService =
+		(*jni_env)->GetMethodID(jni_env,
+			Context, "getSystemService",
+			"(Ljava/lang/String;)Ljava/lang/Object;");
+
+	/* PackageManager local_pkg_mgr = application_context.getPackageManager(); */
+	local_pkg_mgr = 
+		(*jni_env)->CallObjectMethod(jni_env,
+			(*jni)->application_context, Context_getPackageManager);
+
+	(*jni)->package_manager = (*jni_env)->NewGlobalRef(jni_env, local_pkg_mgr);
+
+	/* UsbManager local_usb_mgr =
+  		application_context.getSystemService(Context.USB_SERVICE); */
+	local_usb_mgr =
+		(*jni_env)->CallObjectMethod(jni_env,
+			(*jni)->application_context, Context_getSystemService,
+			Context__USB_SERVICE);
+
+	(*jni)->usb_manager = (*jni_env)->NewGlobalRef(jni_env, local_usb_mgr);
+
+	/* String local_perm_act = "libusb.android.USB_PERMISSION"; */
+	local_perm_act = (*jni_env)->NewStringUTF(jni_env,
+		"libusb.android.USB_PERMISSION");
+
+	(*jni)->permission_action = (*jni_env)->NewGlobalRef(jni_env, local_perm_act);
+
+	(*jni_env)->DeleteLocalRef(jni_env, local_perm_act);
+	(*jni_env)->DeleteLocalRef(jni_env, local_usb_mgr);
+	(*jni_env)->DeleteLocalRef(jni_env, local_pkg_mgr);
+	(*jni_env)->DeleteLocalRef(jni_env, local_context);
+
+	(*jni_env)->DeleteLocalRef(jni_env, Context);
+	(*jni_env)->DeleteLocalRef(jni_env, Context__USB_SERVICE);
+	(*jni_env)->DeleteLocalRef(jni_env, ActivityThread);
+
+	return LIBUSB_SUCCESS;
+}
+
+int android_jni_free(struct android_jni_context *jni)
+{
+	int r;
+	JNIEnv *jni_env;
+
+	r = android_jni_env(jni, &jni_env);
+	if (r != LIBUSB_SUCCESS)
+		free(jni);
+		return r;
```

Comment:
```suggestion
	if (r != LIBUSB_SUCCESS) {
		free(jni);
		return r;
	}
```
what can I say - I love python too

Reply:
thanks =) still have making gpt-neo function locally on my hobby queue (it's a popular python language model)